lib/compress/reduce-vars.js

```diff
@@ -528,93 +539,149 @@ function handle_defined_after_hoist(parent) {
+                    found_symbol_writes.add(found_symbols.length);
                 }
+                found_symbols.push(id);
             }
-        })));
-
-        for (const { def, defun, fname_def } of potential_conflicts) {
-            const defun_range = defun_ranges.get(defun);
-
-            // find the index in `found_symbols`, with some special rules:
-            const find = (sym_id, starting_at = 0, must_be_write = false) => {
-                let index = starting_at;
-
-                for (;;) {
-                    index = found_symbols.indexOf(sym_id, index);
-
-                    if (index === -1) {
-                        break;
-                    } else if (index >= defun_range.start && index < defun_range.end) {
-                        index = defun_range.end;
-                        continue;
-                    } else if (must_be_write && !found_symbol_writes.has(index)) {
-                        index++;
-                        continue;
-                    } else {
-                        break;
-                    }
-                }
+        }
+    })));
+
+    const defun_first_read_map = new Map();
+    for (const defun of defuns_of_interest) {
+        const defun_range = defun_ranges.get(defun);
+        const defun_first_read = found_symbols.findIndex(
+            (sym, index) =>
+                sym === defun &&
+                (index < defun_range.start || index > defun_range.end)
+        );
+        if (defun_first_read === -1) {
+            continue;
+        }
+        defun_first_read_map.set(defun, defun_first_read);
+
+        // Update all depdencies of `defun`
+        let queue = enclosing_defuns_map.get(defun);
+        queue = queue ? queue.slice() : [];
+        const visited = new Set();
+        while (queue.length) {
+            const enclosed_defun = queue.pop();
+            // bail out on circular dependencies
+            if (visited.has(enclosed_defun)) {
+                continue;
+            }
+            visited.add(enclosed_defun);
 
-                return index;
-            };
+            let enclosed_defun_first_read = defun_first_read_map.get(enclosed_defun);
+            if (enclosed_defun_first_read == null) enclosed_defun_first_read = Infinity;
+            if (enclosed_defun_first_read < defun_first_read) {
+                continue;
+            }
 
-            const read_defun_at = find(fname_def.id);
-            const wrote_def_at = find(def.id, read_defun_at + 1, true);
+            defun_first_read_map.set(enclosed_defun, defun_first_read);
 
-            const wrote_def_after_reading_defun = read_defun_at != -1 && wrote_def_at != -1 && wrote_def_at > read_defun_at;
+            for (const enclosed_enclosed_defun of enclosing_defuns_map.get(enclosed_defun) || []) {
+                queue.push(enclosed_enclosed_defun);
+            }
+        }
+    }
+
+    for (const [defun, defs] of defun_enclosing_defs_map) {
```

Comment:
```suggestion
    // ensure write-then-read order, otherwise clear `fixed`
    for (const [defun, defs] of defun_enclosing_defs_map) {
```

Just a small comment to clarify